fix (refactor humanSvg component):

sushma-ananda opened the pull request:
1) move tooltip methods to separate utils
2) add new utils to: HumanSvgComponent.js and Search.js
3) don't get 10 elements by ids from svg: instead re-use pathTriggers
4) remove reduce/maps over already existing arrays
5) have only single loop over pathtriggers
6) remove switch/case over tissues (10 cases): reuse pathtriggers

mccalluc commented on hubmap/frontend/src/commons/human-svg-utils.js:
```diff
@@ -0,0 +1,39 @@
+  tooltipText.firstChild.data = lines[0];
+  updateText('tspanid', lines[1]);
+  const length = tooltipText.getComputedTextLength();
+  for (let i = 0; i < tooltipRects.length; i++) {
```
maybe
```
tooltipRects.forEach((rect) => {
...
});
```

sushma-ananda replied:
it turned out at any given point of time: i.e. when any given mouse pointer move event, tooltipRectangle is always one. because for getElementsByTagName/Classname we always get HTML collection, we have a collection. Using Array.from(HTMLCollection object) was good, but did not require this loop, since, as per lot of tests, there is only 1 element in the Collection all the time. So just get [0] and no need to even loop. i.e. to say, there is always only single tooltip present at any given mouse pointer on screen.

mccalluc commented on hubmap/frontend/src/commons/human-svg-utils.js:
```diff
@@ -0,0 +1,39 @@
+  const humanSvg = document.getElementById('human');
+  const tooltip = humanSvg.getElementById('tooltip');
+  tooltip.setAttributeNS(null, 'visibility', 'hidden');
+  this.setAttribute('opacity', 1);
```
Here and above, I'm confused seeing `this` in a function that isn't a method.

sushma-ananda replied:
yes :/ bad mistake.

mccalluc commented on hubmap/frontend/src/commons/human-svg-utils.js:
```diff
@@ -0,0 +1,39 @@
+  for (let i = 0; i < tooltipRects.length; i++) {
+    tooltipRects[i].setAttributeNS(null, 'width', length + 10);
+    tooltipRects[i].setAttributeNS(null, 'height', 30);
+    tooltipRects[i].setAttributeNS(null, 'font-size', 2);
```
Having a util function to set attributes from an Object would make this more readable. Something like this on SO: https://stackoverflow.com/questions/12274748/setting-multiple-attributes-for-an-element-at-once-with-javascript 
```
function setAttributes(el, attrs) {
  for(var key in attrs) {
    el.setAttribute(key, attrs[key]);
  }
}
```

mccalluc commented on hubmap/frontend/src/components/home/HumanSvgComponent.js:
```diff
@@ -98,140 +58,23 @@ class HumanAnatomyCard extends React.Component {
+			  } else if (pathTriggers[i].id.toLowerCase() === 'main') {
+				 pathTriggers[i].setAttributeNS(null, 'data-tooltip-text', ` # experiments:${(this.experimentState.response.length - 1).toString()}, 9 genes`);
+			  }
+		  });
```
With all the code that's being deleted, you can probably remove most of the eslint-disables at the top.

mccalluc commented on hubmap/frontend/src/commons/human-svg-utils.js:
```diff
@@ -0,0 +1,39 @@
+
+export function hideToolTip() {
+  const humanSvg = document.getElementById('human');
+  const tooltip = humanSvg.getElementById('tooltip');
```
IDs are global: Just go to this directly, I think.

mccalluc commented on hubmap/frontend/src/components/home/HumanSvgComponent.js:
```diff
@@ -98,140 +58,23 @@ class HumanAnatomyCard extends React.Component {
-	    });
-	    human = document.getElementById('main');
-	    human.setAttributeNS(null, 'data-tooltip-text', ` # experiments:${(this.experimentState.response.length - 1).toString()}, 9 genes`);
 	    for (let i = 0; i < pathTriggers.length; i++) {
```
forEach

mccalluc commented on hubmap/frontend/src/components/home/HumanSvgComponent.js:
```diff
@@ -98,140 +58,23 @@ class HumanAnatomyCard extends React.Component {
-	      });
+	      pathTriggers[i].addEventListener('mousemove', utils.showToolTip);
+	      pathTriggers[i].addEventListener('mouseout', utils.hideToolTip);
+		  pathTriggers[i].setAttributeNS(null, 'data-tooltip-text', ` ${pathTriggers[i].id},No gene data available`);
```
Spacing?
```suggestion
		  pathTriggers[i].setAttributeNS(null, 'data-tooltip-text', `${pathTriggers[i].id}: No gene data available`);
```

mccalluc commented on hubmap/frontend/src/components/home/HumanSvgComponent.js:
```diff
@@ -98,140 +58,23 @@ class HumanAnatomyCard extends React.Component {
+	      pathTriggers[i].addEventListener('mousemove', utils.showToolTip);
+	      pathTriggers[i].addEventListener('mouseout', utils.hideToolTip);
+		  pathTriggers[i].setAttributeNS(null, 'data-tooltip-text', ` ${pathTriggers[i].id},No gene data available`);
+		  this.experimentState.response.slice(0, this.experimentState.response.length - 1).forEach((experiment) => {
```
```suggestion
		  this.experimentState.response.slice(0, -1).forEach((experiment) => {
```
Can you also add a comment explaining why you don't want the last one?

mccalluc commented on hubmap/frontend/src/components/home/HumanSvgComponent.js:
```diff
@@ -98,140 +58,23 @@ class HumanAnatomyCard extends React.Component {
+		  this.experimentState.response.slice(0, this.experimentState.response.length - 1).forEach((experiment) => {
+			  if (pathTriggers[i].id.toLowerCase() === experiment.tissue.name.toLowerCase()) {
+				 pathTriggers[i].setAttributeNS(null, 'data-tooltip-text', ` ${experiment.tissue.name},# genes: ${experiment.genes !== undefined ? experiment.genes.length : 0}`);
+			  } else if (pathTriggers[i].id.toLowerCase() === 'main') {
```
All these IDs are under our control, so can we just have a consistent capitalization?

mccalluc commented on hubmap/frontend/src/components/home/HumanSvgComponent.js:
```diff
@@ -98,140 +58,23 @@ class HumanAnatomyCard extends React.Component {
+		  pathTriggers[i].setAttributeNS(null, 'data-tooltip-text', ` ${pathTriggers[i].id},No gene data available`);
+		  this.experimentState.response.slice(0, this.experimentState.response.length - 1).forEach((experiment) => {
+			  if (pathTriggers[i].id.toLowerCase() === experiment.tissue.name.toLowerCase()) {
+				 pathTriggers[i].setAttributeNS(null, 'data-tooltip-text', ` ${experiment.tissue.name},# genes: ${experiment.genes !== undefined ? experiment.genes.length : 0}`);
```
use the attribute setting util you'll create, and space after comma:
```suggestion
				 pathTriggers[i].setAttributeNS(null, 'data-tooltip-text', ` ${experiment.tissue.name}, # genes: ${experiment.genes !== undefined ? experiment.genes.length : 0}`);
```

sushma-ananda commented on hubmap/frontend/src/commons/human-svg-utils.js:
```diff
@@ -0,0 +1,39 @@
+  const tooltipRects = tooltip.getElementsByTagName('rect');
+  const CTM = humanSvg.getScreenCTM();
+  const mousex = (evt.clientX - CTM.e + 6) / CTM.a;
+  const mousey = (evt.clientY - CTM.f + 20) / CTM.d;
```
We get Human SVG's current transformation matrix. (humanSvg.getCTM method).
And if element has attributes x,y then screen coordinates are of the form ax+e, dy+f.
so find real position of mouse pointer (coordinates) in SVG space i.e. ViewBox
by calculating the inverse. adding 6 and 20 are numbers to maintain sufficient distance between mouse cursor and the tooltip rectangle.

mccalluc commented:
I'll re-review this, unless I should wait? I think we should get better about testing... it wouldn't need to be this PR, and it doesn't need to be 100%, but should make a plan for it.

sushma-ananda commented:
> I'll re-review this, unless I should wait? I think we should get better about testing... it wouldn't need to be this PR, and it doesn't need to be 100%, but should make a plan for it.

hi Chuck. I completed incorporating your review comments. Its ready for your review. thanks, I updated your comments in this ongoing issue: https://github.com/hubmapconsortium/hubmap-data-portal/issues/104 

I moved/quoted testing related discussion to #104 since you mentioned you wanted to have this outside of this PR.
